### PR TITLE
Colored hypergraph

### PR DESCRIPTION
**Pull request by mhajij**
Combintorial complex is a hypergraph X with a rank function rk : X->Z^{>0} s.t. if x<=y then rk(x)<=rk(y). This provides the CC structure with ability to consider incidence matrices between rank i and rank j cells.

In many applications however, the rank function is not strictly needed. The rank function makes CC somewhat expensive to compute (add_cell). Instead, one may define a more general class of complexes by defining a color function c: X->Z^{>0} with no condition as before. Such a complex is strictly more general than CCs and allows for faster insertion time. c can be thought of as a 'color' for hyperedges and one may define incidence between two cells of different colors

Write a class ColorHyperGraph with the basic functionalities.

This class has functionalities that overlap with the CC class so creating it will require refactoring the CC as well. Maybe CC will end up inheriting from it. Up to discussion.

### CODE REVIEW

**Review by ffl096 — CHANGES_REQUESTED** [review]
Some preliminary comments, I didn't check everything yet.

In general there are two issues I have right now:
1. There is lots of duplicated code right now between `ColoredHyperGraph` and `CombinatorialComplex`. In a best case scenario, `CombinatorialComplex` could probably extend from the new `ColoredHyperGraph` and validate the ranking function where appropriate.
   If that is too cumbersome extract more into common helper functions.
3.  `rank` vs. `colors`: We need to unify this across both classes, otherwise we will always have a clash where both names are used.
   Right now this class becomes apparent in the constructor of `ColoredHyperGraph`, because `Cell.rank` is used. Renaming it to `Cell.color` doesn't help either, then we have the same problem in the combinatorial complex class.

**Comment by mhajij** [comment]
> Some preliminary comments, I didn't check everything yet.
> 
> In general there are two issues I have right now:
> 
> 1. There is lots of duplicated code right now between `ColoredHyperGraph` and `CombinatorialComplex`. In a best case scenario, `CombinatorialComplex` could probably extend from the new `ColoredHyperGraph` and validate the ranking function where appropriate.

Yes, this is why I said it has so much overlap and we need to refactor the two classes. I think the main differences will be in the add_cell functions and its related functions. 

I think CC should inherit from a colored hypergraph and overload the methods that are different. This way we avoid this issue.

What do you think?

>    If that is too cumbersome extract more into common helper functions.
> 2. `rank` vs. `colors`: We need to unify this across both classes, otherwise we will always have a clash where both names are used.

I am not sure how to resolve this issue. The term 'rank' is misleading in the colored hypergraph.

>    Right now this class becomes apparent in the constructor of `ColoredHyperGraph`, because `Cell.rank` is used. Renaming it to `Cell.color` doesn't help either, then we have the same problem in the combinatorial complex class.

This is a typo. I can fix that.

File: toponetx/classes/colored_hypergraph.py [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+    """Class for ColoredHyperGraph Complex.
+
+    A Colored Hypergraph (CHG) is a triple CHG = (S, X, c) where:
+    - S is an abstract set of entities,
+    - X a subset of the power set of X, and
+    - c is the a color function that associates for every set x in X a color, a positive integer.
+
+
+
+    A CHG is a generlization of graphs, combintorial complex, hypergraphs, cellular and simplicial complexes.
```
[the remark this answers is not in the document]
Comment (mhajij) [reply toponetx/classes/colored_hypergraph.py:35]: sure that makes sense. I will do a better job explaining this class.

**Comment by mhajij** [comment]
> > Some preliminary comments, I didn't check everything yet.
> > In general there are two issues I have right now:
> > 
> > 1. There is lots of duplicated code right now between `ColoredHyperGraph` and `CombinatorialComplex`. In a best case scenario, `CombinatorialComplex` could probably extend from the new `ColoredHyperGraph` and validate the ranking function where appropriate.
> 
> Yes, this is why I said it has so much overlap and we need to refactor the two classes. I think the main differences will be in the add_cell functions and its related functions.
> 
> I think CC should inherit from a colored hypergraph and overload the methods that are different. This way we avoid this issue.
> 
> What do you think?
> 
> > If that is too cumbersome extract more into common helper functions.
> > 2. `rank` vs. `colors`: We need to unify this across both classes, otherwise we will always have a clash where both names are used.
> 
> I am not sure how to resolve this issue. The term 'rank' is misleading in the colored hypergraph.
> 
> > Right now this class becomes apparent in the constructor of `ColoredHyperGraph`, because `Cell.rank` is used. Renaming it to `Cell.color` doesn't help either, then we have the same problem in the combinatorial complex class.
> 
> This is a typo. I can fix that.

@ffl096 , we will end up probably only have to overload a few functions in the CC class this way, everythign else will be inhereted from the colored HyperGraph.

**Pull request closed without merging**